borrowck: do not suggest to change "&mut self" to "&mut mut self"

birkenfeld opened the pull request:
Matching the snippet string might not be the cleanest, but matching
the AST node instead seems to end in a lot of nested `if let`s, so I
don't know what's better.

Of course it's entirely possible that there is another API altogether
that I just don't know of?

Fixes #31424.

rust-highfive commented:
r? @Aatch

(rust_highfive has picked a reviewer for you, use r? to override)

jseyfried commented:
This would prevent the following correct suggestion:
```
struct Struct;

impl Struct {
    fn foo(self: &mut Self) {
        (&mut self).bar();
        //~^ ERROR cannot borrow immutable argument `self` as mutable
        //~| HELP to make the argument mutable, use `mut` as shown:
        //~|    fn foo(mut self: &mut Self) {
    }
    
    fn bar(&mut self) {}
}
```

It might be a good idea to match the AST node so that we can continue providing the suggestion in the above case. It would be even better if we could provide the above `HELP` message even when `self` is declared as `&mut self` instead of `self: &mut Self`.

That being said, I think it's more important to not provide a bad suggestion in the common case of `&mut self`, so r=me unless you would like to work one of the above improvements.

birkenfeld commented:
I've looked a bit at what would be involved, and it's tricky: the "pat" node I can get only ever spans the "self" (it's the special pseudo node added by `Arg::new_self`). The next parent is the whole Item, and the individual Args have no spans. So it'd be a lot of work for this edge case (which is probably quite obscure, since a `&mut &mut T` is not often necessary).

But I've rebased and added a test, will push that now.

jseyfried commented on src/test/compile-fail/issue-31424.rs:
```diff
@@ -0,0 +1,30 @@
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// forbid-output: &mut mut self
```
Nice -- I didn't know about `forbid-output`

jseyfried commented:
@bors r+

bors commented:
:pushpin: Commit fef8276 has been approved by `jseyfried`

<!-- @bors r=jseyfried fef827672ddb0df866c77239f3fc034ed98de556 -->

bors commented:
:hourglass: Testing commit fef8276 with merge 6dbb0e8...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/9131), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/9168), [auto-linux-32cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-32cross-opt/builds/251), [auto-linux-64-cargotest](http://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/443), [auto-linux-64-cross-armhf](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-armhf/builds/254), [auto-linux-64-cross-armsf](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-armsf/builds/258), [auto-linux-64-cross-freebsd](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/257), [auto-linux-64-cross-netbsd](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-netbsd/builds/257), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/2400), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/9106), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/9092), [auto-linux-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-mir/builds/678), [auto-linux-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/1063), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/9088), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/2299), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/4243), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/9144), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/9153), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/9114), [auto-mac-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/1072), [auto-mac-cross-ios-opt](http://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/256), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/4179), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/4192), [auto-win-gnu-32-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/1076), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/4186), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/4168), [auto-win-msvc-32-cross-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-cross-opt/builds/541), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3389), [auto-win-msvc-64-cargotest](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/446), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/4181), [auto-win-msvc-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-mir/builds/681), [auto-win-msvc-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/1080)
